gluonnlp/model/language_model.py

```diff
@@ -421,3 +423,97 @@ def standard_lstm_lm_1500(dataset_name=None, vocab=None, pretrained=False, ctx=c
         ('45d6df33f35715fb760ec8d18ed567016a897df7', 'awd_lstm_lm_1150_wikitext-2'),
         ('7894a046f8286db0d5d2ed672b60f4f52b4bc3aa', 'awd_lstm_lm_600_wikitext-2')
     ]})
+
+class SampledRNN(Block):
+    """Big language model with LSTMP for inference.
+
+    Parameters
+    ----------
+    vocab_size : int
+        Size of the input vocabulary.
+    embed_size : int
+        Dimension of embedding vectors.
+    hidden_size : int
+        Number of hidden units for LSTMP.
+    num_layers : int
+        Number of LSTMP layers.
+    projection_size : int
+        Number of projection units for LSTMP.
+    num_sampled : int
+        Number of sampled classes for the decoder.
+    dropout : float
+        Dropout rate to use for encoder output.
+
+    """
+    def __init__(self, vocab_size, embed_size, hidden_size, num_layers,
+                 projection_size, dropout=0.0, **kwargs):
+        super(SampledRNN, self).__init__(**kwargs)
+        self._embed_size = embed_size
+        self._hidden_size = hidden_size
+        self._projection_size = projection_size
+        self._num_layers = num_layers
+        self._dropout = dropout
+        self._vocab_size = vocab_size
+
+        with self.name_scope():
+            self.embedding = self._get_embedding()
+            self.encoder = self._get_encoder()
+            self.decoder = self._get_decoder()
+
+    def _get_embedding(self):
+        prefix = 'embedding0_'
+        embedding = nn.Sequential()
+        with embedding.name_scope():
+            embedding.add(nn.Embedding(self._vocab_size, self._embed_size, prefix=prefix))
+            if self._dropout:
+                embedding.add(nn.Dropout(self._dropout))
+        return embedding
+
+    def _get_encoder(self):
+        block = rnn.SequentialRNNCell()
+        with block.name_scope():
+            for _ in range(self._num_layers):
+                block.add(contrib.rnn.LSTMPCell(self._hidden_size, self._projection_size))
```

Comment:
Based on my understanding, the projection layer is applied after the last encoder layer. It would be better to just use LSTM layer and apply a projection layer.